starlite/middleware/csrf.py

```diff
@@ -50,18 +52,28 @@ async def __call__(self, scope: "Scope", receive: "Receive", send: "Send") -> No
             await self.app(scope, receive, send)
             return
 
-        request: "Request[Any, Any]" = scope["app"].request_class(scope=scope)
+        request: "Request[Any, Any]" = scope["app"].request_class(scope=scope, receive=receive)
+        content_type, _ = request.content_type
         csrf_cookie = request.cookies.get(self.config.cookie_name)
         existing_csrf_token = request.headers.get(self.config.header_name)
 
+        if not existing_csrf_token and content_type in {
+            RequestEncodingType.URL_ENCODED,
+            RequestEncodingType.MULTI_PART,
+        }:
+            form = await request.form()
+            existing_csrf_token = form.get("_csrf_token", None)
+
         if request.method in self.config.safe_methods:
-            await self.app(scope, receive, self.create_send_wrapper(send=send, csrf_cookie=csrf_cookie))
+            token = scope["_csrf_token"] = generate_csrf_token(secret=self.config.secret)  # type: ignore
```

Comment:
I just got up and understood what has been bothering me with this code.

`csrf_token` will contain new token on each safe request.

As far as I understand TestClient persists cookies between requests so an easy check would be modifying `test_csrf_form_parsing` test to perform two get requests in a row before parsing response with BS

```python
_ = client.get("/")
response = client.get("/")
```

Reply:
true. What would be the solution in your opinion?

Reply:
Do something like this

```python
csrf_cookie_exists = csrf_cookie is None
if not(csrf_cookie_exists):
    token = scope["_csrf_token"] = csrf_cookie
else:
    token = scope["_csrf_token"] = generate_csrf_token(secret=self.config.secret)  # type: ignore
await self.app(scope, receive, self.create_send_wrapper(send=send, set_cookie=csrf_cookie_exists, token=token))
.....

def create_send_wrapper(self, send: "Send", token: str, set_cookie: bool) -> "Send":
....

    async def send_wrapper(message: "Message") -> None:
    ....
        if set_cookie and message["type"] == "http.response.start":
            ...
    return send_wrapper

```

Reply:
Sorry, I did this quickly out of my head so the code contains too many `not`. Should rename the variable to `csrf_cookie_doesnot_exists` and negate it so we can omit `not`